File: 2023/01B/main.hs
module Main where

import Data.Bool (bool)
import Data.Char (digitToInt, isDigit)
import Data.Foldable (find)
import Data.Function ((&))
import Data.List (isPrefixOf, tails)
import Data.Maybe (catMaybes, listToMaybe)

type Case = [String]

type Soln = Int

main :: IO ()
main = interact process

process :: String -> String
process = showSoln . solve . readCase

readCase :: String -> Case
readCase = lines

showSoln :: Soln -> String
showSoln = unlines . return . show

solve :: Case -> Soln
solve = sum . map readCalibrationValue

readCalibrationValue :: String -> Int
readCalibrationValue = (\xs -> head xs * 10 + last xs) . readDigits

readDigits :: String -> [Int]
readDigits = catMaybes . map readDigit . tails

readDigit :: String -> Maybe Int
readDigit = listToMaybe . catMaybes . flip map rules . (&)
  where
    rules =
      [ (fmap digitToInt . find isDigit . listToMaybe)
      , (bool Nothing (Just 1) . isPrefixOf "one")
      , (bool Nothing (Just 2) . isPrefixOf "two")
      , (bool Nothing (Just 3) . isPrefixOf "three")
      , (bool Nothing (Just 4) . isPrefixOf "four")
      , (bool Nothing (Just 5) . isPrefixOf "five")
      , (bool Nothing (Just 6) . isPrefixOf "six")
      , (bool Nothing (Just 7) . isPrefixOf "seven")
      , (bool Nothing (Just 8) . isPrefixOf "eight")
      , (bool Nothing (Just 9) . isPrefixOf "nine") ]

File: 2023/01B/main.rs
use std::io;

fn main() {
    let stdin = io::stdin();
    let lines = stdin.lines().collect::<Result<Vec<_>, _>>().unwrap();
    let sum = lines
        .iter()
        .map(|line| {
            // I *should* use find and rfind to search for pattern matches specified by index and
            // value and select the leftmost and rightmost ones...
            // Or I could just hackily modify the string so that it contains the digits now.
            // Bonus ugly because we don't want to break any overlapping patterns.
            let line = line
                .replace("one", "o1e")
                .replace("two", "t2o")
                .replace("three", "t3e")
                .replace("four", "4")
                .replace("five", "5e")
                .replace("six", "6")
                .replace("seven", "7n")
                .replace("eight", "e8t")
                .replace("nine", "n9e");
            let digits: Vec<u32> = line.chars().filter_map(|c| c.to_digit(10)).collect();
            let head = digits.first()?;
            let last = digits.last()?;
            Some(head * 10 + last)
        })
        .sum::<Option<u32>>()
        .unwrap();
    println!("{}", sum);
}
